src/transformers/models/fastspeech2/modeling_fastspeech2.py

```diff
@@ -0,0 +1,697 @@
+
+    def get_energy_embedding(self, hidden, target=None, factor=1.0):
+        out = self.energy_predictor(hidden)
+        bins = self.energy_bins.to(hidden.device)
+        if target is None:
+            out = out * factor
+            energy_embedding = self.embed_energy(torch.bucketize(out, bins))
+        else:
+            energy_embedding = self.embed_energy(torch.bucketize(target, bins))
+        return out, energy_embedding
+
+    def forward(
+        self,
+        hidden,
+        padding_mask,
+        durations=None,
+        pitches=None,
+        energies=None,
+        d_factor=1.0,
+        p_factor=1.0,
+        e_factor=1.0,
+    ):
+        # hidden.shape == (batch_size, seq_length, num_channels)
+        log_dur_out = self.duration_predictor(hidden)
+        dur_out = torch.clamp(torch.round((torch.exp(log_dur_out) - 1) * d_factor).long(), min=0)
+        dur_out.masked_fill_(padding_mask, 0)
+
+        pitch_out, pitch_embedding = self.get_pitch_embedding(hidden, pitches, p_factor)
+        hidden = hidden + pitch_embedding
+        energy_out, energy_embedding = self.get_energy_embedding(hidden, energies, e_factor)
+        hidden = hidden + energy_embedding
+
+        hidden, out_lens = self.length_regulator(hidden, dur_out if durations is None else durations)
+
+        return hidden, out_lens, log_dur_out, pitch_out, energy_out
+
+
+class Postnet(nn.Module):
+    def __init__(self, in_dim, n_channels, kernel_size, n_layers, dropout):
+        super(Postnet, self).__init__()
+        self.convolutions = nn.ModuleList()
+        for i in range(n_layers):
+            cur_layers = [
+                nn.Conv1d(
+                    in_dim if i == 0 else n_channels,
+                    n_channels if i < n_layers - 1 else in_dim,
+                    kernel_size=kernel_size,
+                    padding=((kernel_size - 1) // 2),
+                ),
+                nn.BatchNorm1d(n_channels if i < n_layers - 1 else in_dim),
+            ]
+            if i < n_layers - 1:
+                cur_layers.append(nn.Tanh())
+            cur_layers.append(nn.Dropout(dropout))
+            nn.init.xavier_uniform_(
+                cur_layers[0].weight,
+                nn.init.calculate_gain("tanh" if i < n_layers - 1 else "linear"),
+            )
+            self.convolutions.append(nn.Sequential(*cur_layers))
+
+    def forward(self, hidden):
+        # hidden.shape == (batch_size, seq_length, num_channels)
+        hidden = hidden.transpose(1, 2)
+        for conv in self.convolutions:
+            hidden = conv(hidden)
+        return hidden.transpose(1, 2)
+
+
+class FastSpeech2Encoder(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.padding_idx = config.pad_token_id
+        self.n_frames_per_step = config.n_frames_per_step
+        self.out_dim = config.output_frame_dim * config.n_frames_per_step
+
+        self.spk_emb_proj = None
```

Comment:
Let's remove this no?

Reply:
Why should we set a weight to None?

Reply:
Multi-speaker checkpoints (e.g. common voice) will have a speaker embedding, but a single-speaker checkpoint like LJSpeech won't. Do you think there is perhaps a better way to deal with this?